mlbench_core/optim/pytorch/optim.py

```diff
@@ -499,12 +499,290 @@ def step(self, closure=None):
+            torch.matmul(matrix, q, out=p)
+
+        all_reduce(self.p_memory)
+
+        # Start communicating rank 1 tensors
+        rank1_tensor_list = TensorBuffer([tensor for (tensor, _, _) in rank1_tensors])
+
+        rank1_handle = rank1_tensor_list.all_reduce(async_op=True)
+
+        for p in ps:
+            orthogonalize(p)
+
+        for p, q, (tensor, _, _) in zip(ps, qs, high_rank_tensors):
+            matrix = tensor.view(tensor.shape[0], -1)
+            torch.matmul(matrix.t(), p, out=q)
+
+        all_reduce(self.q_memory)
+        self.q_memory.data[:] /= self.n_workers
+
+        for p, q, (tensor, out, mem) in zip(ps, qs, high_rank_tensors):
+            # Set the output gradient
+            torch.matmul(p, q.t(), out=out.data[:])
+            mem.data[:] = tensor - out
+
+        rank1_handle.wait()
+        rank1_tensor_list.buffer /= self.n_workers
+        rank1_tensor_list.unpack([out for (_, out, _) in rank1_tensors])
+
+
+class TensorBuffer:
+    """
+    Packs multiple tensors into one flat buffer for efficient
+    intra-worker communication.
+    """
+
+    def __init__(self, tensors):
+        indices = [0]
+        for tensor in tensors:
+            new_end = indices[-1] + tensor.nelement()
+            indices.append(new_end)
+
+        self._start_idx = indices[:-1]
+        self._end_idx = indices[1:]
+        self._tensors = tensors
+
+        self.buffer = torch.cat([t.view(-1) for t in tensors])  # copies
+
+    def __getitem__(self, index):
+        return self.buffer[self._start_idx[index]: self._end_idx[index]].view(*self._tensors[index].shape)
+
+    def __len__(self):
+        return len(self._tensors)
+
+    def pack(self, tensors=None):
+        # Optional. init already does this.
+        if tensors is None:
+            tensors = self._tensors
+        for tensor, entry in zip(tensors, self):
+            entry[:] = tensor
+
+    def unpack(self, tensors):
+        for tensor, entry in zip(tensors, self):
+            tensor[:] = entry
+
+    def nelement(self):
+        return self.buffer.nelement()
+
+    def element_size(self):
+        return self.buffer.element_size()
+
+    def bits(self):
+        return 8 * self.nelement() * self.element_size()
+
+    def all_reduce(self, async_op=False):
+        return torch.distributed.all_reduce(self.buffer, async_op=async_op)
+
+    def all_gather(self, async_op=False):
+        n_workers = torch.distributed.get_world_size() if torch.distributed.is_available() else 1
+        buffers = [torch.empty_like(self.buffer) for i in range(n_workers)]
+        handle = all_gather(buffers, self.buffer, async_op=async_op)
+        if async_op:
+            return buffers, handle
+        else:
+            return buffers
+
+
+def all_reduce(*args, **kwargs):
+    if torch.distributed.is_available() and torch.distributed.get_world_size() > 1:
+        return torch.distributed.all_reduce(*args, **kwargs)
+
+
+@torch.jit.script
+def orthogonalize(matrix):
```

Comment:
Maybe move this to a utility module ? Maybe create a file `mlbench_core/optim/pytorch/utils.py`, as we will probably need it later.